### PR TITLE
Improve performance of `JoinParts` for spans

### PR DESCRIPTION
**Pull request by Sewer56**
This improves the performance of the `JoinParts` method, by avoiding an unnecessary heap allocation which makes another copy of the string.

We pin the Span (in the case it originated from the heap), and use the pointers to the characters as input to `string.Create`.

### CODE REVIEW

File: src/NexusMods.Paths/Utilities/PathHelpers.cs [hunk 1]
```diff
@@ -386,15 +386,24 @@ public static string JoinParts(ReadOnlySpan<char> left, ReadOnlySpan<char> right
...
-        if (count == 0) return string.Empty;
-
-        Debug.Assert(count == spanLength, $"Calculated span length '{spanLength}' doesn't match actual span length '{count}'");
-        return buffer.ToString();
+        unsafe
+        {
+            fixed (char* leftPtr = left)
+            fixed (char* rightPtr = right)
+            {
+                return string.Create(spanLength, (((IntPtr)leftPtr, left.Length), ((IntPtr)rightPtr, right.Length), os), (span, tuple) =>
```
Comment (erri120) [inline src/NexusMods.Paths/Utilities/PathHelpers.cs:394]: The `Debug.Assert` call references `spanLength` so there's still a delegate allocation. This is also an issue with the string-counterpart of `JoinParts`.

Comment (Sewer56) [reply src/NexusMods.Paths/Utilities/PathHelpers.cs:394]: Yup. I don't consider it a big deal though. In release builds, where the `DEBUG` preprocessor definition isn't enabled, this code will be eliminated.

**Review by erri120 — CHANGES_REQUESTED** [review]
This needs some benchmarks to compare old vs new. The span-version of `JoinParts` is also not used by anything except the tests, so even if the new version is more performant that the old version, we won't see that benefit.

If I remember correctly, we have a span-version and a string-version of `JoinParts` because they have different performance characteristics. The string-version uses `string.Create` to get a `Span<char>` while the span-version used to allocate an array "the normal way". Due to ref-struct weirdness, I couldn't get the span-version to use `string.Create`. This PR circumvents the ref-struct weirdness by deconstructing and reconstructing the span inside the `string.Create` part.

With this PR, the span- and string-versions should have similar performance characteristics (needs to be benchmarked), so the string-version should be removed.

You should also add a _short_ comment on the unsafe part explaining what I mentioned above about the span deconstructing and reconstructing inside the `stirng.Create` part.

**Comment by Sewer56** [comment]
```
| Method    | Path1                | Path2                | Mean     | Error    | StdDev   | Gen0   | Allocated |
|---------- |--------------------- |--------------------- |---------:|---------:|---------:|-------:|----------:|
| Current   | a/ver(...)forms [46] | a/ver(...)forms [46] | 20.98 ns | 0.150 ns | 0.141 ns | 0.0124 |     208 B |
| OldMethod | a/ver(...)forms [46] | a/ver(...)forms [46] | 27.12 ns | 0.370 ns | 0.346 ns | 0.0124 |     208 B |
| Current   | a/ver(...)forms [46] | short                | 18.00 ns | 0.188 ns | 0.167 ns | 0.0076 |     128 B |
| OldMethod | a/ver(...)forms [46] | short                | 23.57 ns | 0.501 ns | 0.536 ns | 0.0076 |     128 B |
| Current   | short                | a/ver(...)forms [46] | 17.91 ns | 0.081 ns | 0.063 ns | 0.0076 |     128 B |
| OldMethod | short                | a/ver(...)forms [46] | 24.16 ns | 0.447 ns | 0.349 ns | 0.0076 |     128 B |
| Current   | short                | short                | 14.93 ns | 0.347 ns | 0.399 ns | 0.0029 |      48 B |
| OldMethod | short                | short                | 19.71 ns | 0.084 ns | 0.070 ns | 0.0029 |      48 B |
```

Sure you can have a benchmark if you want.
I just knew it would be faster from a lot of past experience optimizing down to this level.

**Comment by erri120** [comment]
The benchmark doesn't compare the old span-version to the new span-version. It compares the old span-version to the unchanged string-version.

**Comment by Sewer56** [comment]
Ah, apologies. Am sleepy and haven't even noticed, one moment.

**Comment by erri120** [comment]
```
| Method    | Path1                | Path2                | Mean     | Error    | StdDev   | Gen0   | Allocated |
|---------- |--------------------- |--------------------- |---------:|---------:|---------:|-------:|----------:|
| Current   | a/ver(...)forms [46] | a/ver(...)forms [46] | 38.11 ns | 0.326 ns | 0.289 ns | 0.0124 |     208 B |
| OldMethod | a/ver(...)forms [46] | a/ver(...)forms [46] | 29.69 ns | 0.634 ns | 1.190 ns | 0.0124 |     208 B |
| Current   | a/ver(...)forms [46] | short                | 35.46 ns | 0.666 ns | 0.866 ns | 0.0076 |     128 B |
| OldMethod | a/ver(...)forms [46] | short                | 24.90 ns | 0.536 ns | 0.677 ns | 0.0076 |     128 B |
| Current   | short                | a/ver(...)forms [46] | 36.42 ns | 0.750 ns | 0.833 ns | 0.0076 |     128 B |
| OldMethod | short                | a/ver(...)forms [46] | 27.04 ns | 0.585 ns | 1.070 ns | 0.0076 |     128 B |
| Current   | short                | short                | 31.94 ns | 0.657 ns | 0.703 ns | 0.0029 |      48 B |
| OldMethod | short                | short                | 19.94 ns | 0.270 ns | 0.226 ns | 0.0029 |      48 B |
```

This is the results comparing the old span-version to the new span-version.

**Comment by Sewer56** [comment]
```
| Method    | Path1                | Path2                | Mean     | Error    | StdDev   | Gen0   | Allocated |
|---------- |--------------------- |--------------------- |---------:|---------:|---------:|-------:|----------:|
| Current   | a/ver(...)forms [46] | a/ver(...)forms [46] | 17.77 ns | 0.188 ns | 0.157 ns | 0.0124 |     208 B |
| OldMethod | a/ver(...)forms [46] | a/ver(...)forms [46] | 26.01 ns | 0.205 ns | 0.192 ns | 0.0124 |     208 B |
| Current   | a/ver(...)forms [46] | short                | 15.58 ns | 0.179 ns | 0.168 ns | 0.0076 |     128 B |
| OldMethod | a/ver(...)forms [46] | short                | 22.75 ns | 0.187 ns | 0.156 ns | 0.0076 |     128 B |
| Current   | short                | a/ver(...)forms [46] | 15.43 ns | 0.069 ns | 0.054 ns | 0.0076 |     128 B |
| OldMethod | short                | a/ver(...)forms [46] | 21.63 ns | 0.197 ns | 0.185 ns | 0.0076 |     128 B |
| Current   | short                | short                | 12.94 ns | 0.152 ns | 0.142 ns | 0.0029 |      48 B |
| OldMethod | short                | short                | 15.75 ns | 0.104 ns | 0.087 ns | 0.0029 |      48 B |
```

I forgot to account for the fact that nested tuples with deconstruction account for some very unoptimal codegen.
In any case, the core idea wasn't wrong. I fixed the PR.

File: src/NexusMods.Paths/Utilities/PathHelpers.cs [hunk 2]
```diff
@@ -388,30 +390,38 @@ public static string JoinParts(ReadOnlySpan<char> left, ReadOnlySpan<char> right
...
+            });
         }
     }
 
+    unsafe struct JoinPartsParams
+    {
+        internal ReadOnlySpan<char>* Left;
+        internal ReadOnlySpan<char>* Right;
+        internal IOSInformation Os;
+    }
```
Comment (erri120) [inline src/NexusMods.Paths/Utilities/PathHelpers.cs:422]: The struct should be marked as `private` as well as `readonly`.

Comment (Sewer56) [reply src/NexusMods.Paths/Utilities/PathHelpers.cs:422]: I'll do it when I get back to my machine, dw

**Comment by erri120** [comment]
The current string-version is still faster than the new span-version, and all of our code uses the string-version. We can probably just remove the span-version of `JoinParts`.

```
Current = span-version
OldMethod = string-version

| Method    | Path1                | Path2                | Mean     | Error    | StdDev   | Gen0   | Allocated |
|---------- |--------------------- |--------------------- |---------:|---------:|---------:|-------:|----------:|
| Current   | a/ver(...)forms [46] | a/ver(...)forms [46] | 16.75 ns | 0.031 ns | 0.024 ns | 0.0124 |     208 B |
| OldMethod | a/ver(...)forms [46] | a/ver(...)forms [46] | 15.86 ns | 0.062 ns | 0.058 ns | 0.0124 |     208 B |
| Current   | a/ver(...)forms [46] | short                | 15.38 ns | 0.348 ns | 0.542 ns | 0.0076 |     128 B |
| OldMethod | a/ver(...)forms [46] | short                | 14.70 ns | 0.335 ns | 0.314 ns | 0.0076 |     128 B |
| Current   | short                | a/ver(...)forms [46] | 15.08 ns | 0.126 ns | 0.118 ns | 0.0076 |     128 B |
| OldMethod | short                | a/ver(...)forms [46] | 14.11 ns | 0.113 ns | 0.106 ns | 0.0076 |     128 B |
| Current   | short                | short                | 12.44 ns | 0.077 ns | 0.072 ns | 0.0029 |      48 B |
| OldMethod | short                | short                | 11.60 ns | 0.062 ns | 0.052 ns | 0.0029 |      48 B |
```

**Comment by Sewer56** [comment]
> The current string-version is still faster than the new span-version.

They're the same code, only catch is there's an additional 3 stack copies and 2 derererences in the Span version, which accounts for a nanosecond. They're unfortunately unavoidable, for now.

If they use a throwhelper inside `string.Create` in a future runtime, making it inlineable, and the `Action<T>` also got inlined, the JIT would then turn the reference into direct variable use and the difference would be 0. Hopefully in a few runtimes from now, hahaha.

**Review by erri120 — APPROVED** [review]

**Pull request merged**